### PR TITLE
Add `coin::split_n` function

### PR DESCRIPTION
**Pull request by aschran**
Splits a `Coin` into `n` (best-effort) equal parts

### CODE REVIEW

**Comment by aschran** [comment]
Is it worth adding an assert that `balance::value(&self.balance) >= n`?

File: crates/sui-framework/sources/coin.move [hunk 1]
```diff
@@ -224,6 +224,17 @@ module sui::coin {
         )
     }
 
+    /// Split coin `self` into `n` coins with equal balances, or as close as
+    /// possible, if the balance is not evenly divisible by `n`.
+    public entry fun split_n<T>(self: &mut Coin<T>, n: u64, ctx: &mut TxContext) {
```
Comment (sblackshear) [inline crates/sui-framework/sources/coin.move:229]: For `n = 0`, I think we either want to fail with an appropriate error-code (preferred) or no-op. Otherwise, we'll fail with an underflow error at 232.

File: crates/sui-framework/sources/coin.move [hunk 2]
```diff
@@ -224,6 +224,17 @@ module sui::coin {
         )
     }
 
+    /// Split coin `self` into `n` coins with equal balances, or as close as
+    /// possible, if the balance is not evenly divisible by `n`.
```
Comment (sblackshear) [inline crates/sui-framework/sources/coin.move:228]: I think it is useful to document where the remainder ends up in the "not evenly divisible" case (I think the answer is "in `self`"?).

File: crates/sui-framework/sources/coin.move [hunk 3]
```diff
@@ -224,6 +224,17 @@ module sui::coin {
...
+    public entry fun split_n<T>(self: &mut Coin<T>, n: u64, ctx: &mut TxContext) {
+        let i = 0;
+        let split_amount = balance::value(&self.balance) / n;
+        while (i < n - 1) {
+            split(self, split_amount, ctx);
+            i = i + 1;
+        };
+    }
```
Comment (sblackshear) [inline crates/sui-framework/sources/coin.move:236]: I think a useful, related function would be something like

```
public fun split_n<T>(self: &mut Coin<T>, n: u64, ctx: &mut TxContext): vector<Coin<T>>
```

, since this version allows the caller to continue using the split pieces instead of immediately transferring them.

Might make sense to have both API's and implement one in terms of the other.

File: crates/sui-framework/sources/coin.move (hunk 2 above)
Comment (aschran) [reply crates/sui-framework/sources/coin.move:228]: Done.

File: crates/sui-framework/sources/coin.move (hunk 3 above)
Comment (aschran) [reply crates/sui-framework/sources/coin.move:260]: PS open to suggestions for what to name the new variant :)

File: crates/sui-framework/sources/coin.move [hunk 4]
```diff
@@ -224,6 +230,35 @@ module sui::coin {
...
+        vec
+    }
+
+    /// Split coin `self` into `n` coins with equal balances. If the balance is
+    /// not evenly divisible by `n`, the remainder is left in `self`.
+    public entry fun split_n<T>(self: &mut Coin<T>, n: u64, ctx: &mut TxContext) {
+        let vec: vector<Coin<T>> = split_n_return(self, n, ctx);
+        let i = 0;
+        let len = vector::length(&vec);
+        while (i < len) {
```
Comment (aschran) [inline crates/sui-framework/sources/coin.move:255]: I could also simplify this a bit by calling `vector::length` in the `while` loop directly, but I wasn't sure if it is at all worth avoiding repeatedly calling that function.

**Comment by aschran** [comment]
Sorry if I'm review spamming anyone. Still getting used to Github PR workflows and I think I maybe should have merged instead of rebased, or something like that 😬

**Comment by randall-Mysten** [comment]
> Sorry if I'm review spamming anyone. Still getting used to Github PR workflows and I think I maybe should have merged instead of rebased, or something like that 😬

All good, not spamming. Rebase and merge should essentially result in the same

> Sorry if I'm review spamming anyone. Still getting used to Github PR workflows and I think I maybe should have merged instead of rebased, or something like that 😬

No worries, we were all new at one point :)

File: crates/sui-framework/sources/coin.move (hunk 3 above)
Comment (awelc) [reply crates/sui-framework/sources/coin.move:260]: > PS open to suggestions for what to name the new variant :)

I'd say that `split_n_return`'s signature already implies that the split coins are returned so I'd change it to simply `split_n` and have the other one called `split_n_transfer` as it's unclear from signature what happens inside beyond splitting.

Comment (aschran) [reply crates/sui-framework/sources/coin.move:260]: I thought about that but was worried it would be confusing because `split` and `split_vec` both transfer without having `transfer` in their name. So if this one has a `_transfer` suffix then it might make it seem like the others *don't* transfer? What do you think?

Comment (awelc) [reply crates/sui-framework/sources/coin.move:260]: Good point. It's probably more consistent to leave it as is in this case. It bothers me a bit that `split_n_return` is somewhat redundant, but I am not super religious about these things :-)

Comment (aschran) [reply crates/sui-framework/sources/coin.move:260]: No argument here that the name isn't ideal! Maybe `split_n_into_vec`? That gets a bit wordy, but at least it's clear.

Comment (Clay-Mysten) [reply crates/sui-framework/sources/coin.move:260]: > split_n_into_vec

Perhaps save a syllable and make it `split_n_to_vec` with the "in" implied?

Comment (aschran) [reply crates/sui-framework/sources/coin.move:260]: I like it, done!

**Review by sblackshear — APPROVED** [review]
LGTM! Please add a separate unit test for `split_n_to_vec`, but otherwise good to go.
